### ISSUE TITLE
Need help: MQTT Client with Sparkfun AssetTracker (SARA-R5) APIs

### ISSUE THREAD

**Issue by Davefries1966** [sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library#34]
Hi all, 
I purchased a Sparkfun AssetTracker development kit with MicroMod ESP32 add-on, GNSS antenna and so on.

I’m wondering how to find examples about using MQTT (I would use the onboard LTE modem instead of the Wi-Fi connection, so no "PubSubClient" library ....).

I’ve looked at AssetTracker APIs and tested some examples, all working with my environment (Arduino IDE on Windows 11), but I didn’t find how to correctly use the “MQTT” APIs.

I mean, I’ve written some code which seems to be able to connect (i.e., the **connectMQTT()** method returns OK) to some public MQTT brokers (e.g., MOSQUITTO, EQMX, etc) but:

- Publish and subscribe methods return errors.
- I didn’t actually understand how/where to use the **readMQTT()** method.
- I set the command callback, but I don’t understand its purpose (I would use **readMQTT()** instead to read topic contents).
- How to set a callback to read incoming topics which I’ve subscribed to ?

Please find attached below the code I've written, just to understand where I’m doing wrong...

Thanks in advance for your support! :-)

Best regards
David

[MQTT_example.txt](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/files/13006307/MQTT_example.txt)

**Comment by PaulZC** [comment]
Hi David (@Davefries1966 ),

As you've found, we don't have a ready-to-go MQTT example for the SARA-R5. We should add one...

Things that will help are:

* The [PointPerfect MQTT example](https://github.com/sparkfun/SparkFun_u-blox_GNSS_v3/tree/main/examples/ZED-F9P/Example18_PointPerfectClient) from the u-blox GNSS library. This uses WiFi to establish the MQTT connection, so not directly relevant, but it does walk nicely through the steps needed.

* [LTE.h](https://github.com/mazgch/hpg/blob/main/software/LTE.h) from @mazgch 's fantastic [HPG Solution](https://github.com/mazgch/hpg/tree/main). This does use the SARA-R5 / LARA-R6 and this library to establish the MQTT connection.

I'm leaning towards writing a u-blox GNSS Assist-Now-via-MQTT example, replicating [Example12](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/tree/main/examples/SARA-R5_Example12_AssistNowOnline) but using MQTT instead of HTTP. The topic details are [here](https://developer.thingstream.io/guides/location-services/pointperfect-getting-started#h.vn9iuugch1ah).

It will take me a while to write the example as I have many other tasks I need to complete first.

If you're in a hurry, please work your way through the HPG Solution code. The answer is in there!

Best wishes,
Paul

**Comment by Davefries1966** [comment]
Thank you Paul for your fast & precious support! :-)
I will work on it in the mainwhile as per your suggestions, and look
forward to seeing further integrations.

Have a great life!

Cheers
David

Il mer 18 ott 2023, 12:38 Paul ***@***.***> ha scritto:

**Comment by Davefries1966** [comment]
Hi, @PaulZC 
I am trying the HPG example and started learning how to use your AssetTracker APIs.... both are fantastic softwares indeed (and thank you again so much for the suggestion!).
However I am stuck with MQTT connect, I mean, the call to **connectMQTT()** method returns 0 (SARA_R5_SUCCESS) but the next call to **getMQTTprotocolError()** returns **error = 13 and errro2 = 34**. I am not able to understand those error codes, anyway the state machine of HPG software doesn't advance to "MQTT" state.

Any suggestion would be so much appreciated! -:)

P.S. Another issue: In my software example, I defined a callback for MQTT commands (through **setMQTTCommandCallback()**), but it is never invoked the one in HPG software...

**Comment by PaulZC** [comment]
Hi David,

The error codes are hidden in the SARA-R5 AT Command Reference: https://www.u-blox.com/sites/default/files/SARA-R5_ATCommands_UBX-19047455.pdf . Check Section 30 for MQTT and Section A.8.4 for the MQTT error codes.

13 is an MQTT error (from A8). 34 is No Network Service (from A.8.4).

Are you sure you have configured the PSD (Packet Switched Data) connection correctly? Maybe try a [Ping](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/blob/main/examples/SARA-R5_Example8_Ping/SARA-R5_Example8_Ping.ino) test, just to be sure?

Hope this helps,
Paul

**Comment by Davefries1966** [comment]
I just ran the Ping example, I inserted "www.google.it" as destination address and didn't get any response...this sounds very strange because the operator seems to be active (VODAFONE ITALY)...for example, the SMS example runs fine.

11:29:11.337 -> SARA-R5 connected!
11:29:11.337 -> 
11:29:11.378 -> Connected to: 22210
11:29:11.421 -> Warning: performPDPaction (deactivate profile) failed. Probably because no profile was active.
11:29:11.934 -> 
11:29:11.934 -> *** Set the Serial Monitor line ending to Newline ***
11:29:11.934 -> 
11:29:11.934 -> Enter the name of the server you want to ping (followed by LF / Newline):

**Comment by Davefries1966** [comment]
This is the result instead with 8.8.8.8:

11:32:51.341 -> SARA-R5 Example
11:32:51.341 -> Press any key to begin
11:32:53.897 -> SARA-R5 connected!
11:32:53.897 -> 
11:32:53.935 -> Connected to: 22210
11:32:53.970 -> Warning: performPDPaction (deactivate profile) failed. Probably because no profile was active.
11:32:54.932 -> 
11:32:54.932 -> *** Set the Serial Monitor line ending to Newline ***
11:32:54.932 -> 
11:32:54.932 -> Enter the name of the server you want to ping (followed by LF / Newline): 
11:33:05.921 -> 
11:33:05.921 -> Ping Result:  Retry #:1  Ping Size (Bytes):32  Remote Host:""  IP Address:"8.8.8.8"  Time To Live (hops):0  Round Trip (ms):-1
11:33:10.943 -> 
11:33:10.943 -> Ping Result:  Retry #:2  Ping Size (Bytes):32  Remote Host:""  IP Address:"8.8.8.8"  Time To Live (hops):0  Round Trip (ms):-1
11:33:15.917 -> 
11:33:15.917 -> Ping Result:  Retry #:3  Ping Size (Bytes):32  Remote Host:""  IP Address:"8.8.8.8"  Time To Live (hops):0  Round Trip (ms):-1
11:33:20.949 -> 
11:33:20.949 -> Ping Result:  Retry #:4  Ping Size (Bytes):32  Remote Host:""  IP Address:"8.8.8.8"  Time To Live (hops):0  Round Trip (ms):-1

**Comment by PaulZC** [comment]
Check you are following the same steps as [Example4](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/tree/main/examples/SARA-R5_Example4_RegisterOperator) and [Example7](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/tree/main/examples/SARA-R5_Example7_ConfigurePacketSwitchedData).

You may need to change:

```
const mobile_network_operator_t MOBILE_NETWORK_OPERATOR = MNO_GLOBAL;
```

Try ```MNO_STD_EUROPE``` or ```MNO_SIM_ICCID```

**Comment by Davefries1966** [comment]
ok...the ping example seems to work only with IP addresses and not with URLs, this is the result with the IP of "test.mosquitto.org":

11:38:51.467 -> Ping Result:  Retry #:1  Ping Size (Bytes):32  Remote Host:""  IP Address:"91.121.93.94"  Time To Live (hops):0  Round Trip (ms):-1
11:38:56.431 -> 
11:38:56.431 -> Ping Result:  Retry #:2  Ping Size (Bytes):32  Remote Host:""  IP Address:"91.121.93.94"  Time To Live (hops):0  Round Trip (ms):-1
11:39:01.455 -> 
11:39:01.455 -> Ping Result:  Retry #:3  Ping Size (Bytes):32  Remote Host:""  IP Address:"91.121.93.94"  Time To Live (hops):0  Round Trip (ms):-1
11:39:06.436 -> 
11:39:06.436 -> Ping Result:  Retry #:4  Ping Size (Bytes):32  Remote Host:""  IP Address:"91.121.93.94"  Time To Live (hops):0  Round Trip (ms):-1

**Comment by Davefries1966** [comment]
I missed to run ALL the examples, they are very inspiring... this is the output of the Example nr. 4: actually the operator selection is ok, but the actual connection NOT; maybe the SIM (which is not for IoT) is not enabled for data connection outside the cell phone (or maybe I have to enable it before....):

11:45:48.810 -> Press any key scan for networks..
11:45:54.465 -> Scanning for networks...this may take up to 3 minutes
11:45:54.465 -> 
11:47:59.099 -> Found 1 operators:
11:47:59.099 -> 1: vodafone IT (22210) - AVAILABLE - LTE
11:47:59.099 -> 2: use automatic selection
11:47:59.099 -> 
11:47:59.099 -> Press 1-2 to select an operator.
11:48:07.875 -> Connecting to option 1
11:48:29.903 -> Error connecting to operator. Reset and try again, or try another network.
11:48:29.903 -> Connection info:
11:48:29.903 -> Context ID:	APN Name:	IP Address:
11:48:31.275 -> RSSI: 18
11:48:31.275 ->

**Comment by Davefries1966** [comment]
If I remember well, with m-center software the connection was OK.... I have to try again...

**Comment by Davefries1966** [comment]
I’m beginning to have doubts that the SIM is not enabled for LTE, but only for 4G.... my fault, sorry! :-(

**Comment by PaulZC** [comment]
That's OK - no worries! I'm in the UK and - in my area - only O2 (Telefonica) provides NB-IoT connectivity. It took me a while to discover that = and get the right SIM...!

I'll leave this open until we have our MQTT example ready. If your code can be turned into a 'generic' Arduino code example, please send us a Pull Request and we'll add it. We appreciate it.

Best wishes,
Paul

**Comment by Davefries1966** [comment]
Very decent of you Paul, I really appreciate your work ! :-)

Cheers
David

**Comment by Davefries1966** [comment]
Hi @PaulZC 

just for the avoidance of any doubt, an usual 4G+ SIM would be suitable for data connection with Sparkfun SARA R5 modem ?
Or, must I purchase a dedicated SIM for IoT ?

Thanks!
David

**Comment by Davefries1966** [comment]
This is the result with examples (register operator + PDP activation + ping).... no PING result (PING callback is not invoked):

11:46:55.170 -> Setting mobile-network operator
11:46:55.239 -> Set mobile network operator to standard Europe
11:46:55.239 -> 
11:46:55.239 -> Press any key scan for networks..
11:46:57.641 -> Scanning for networks...this may take up to 3 minutes
11:46:57.641 -> 
11:47:15.392 -> Found 1 operators:
11:47:15.392 -> 1: vodafone IT (22210) - AVAILABLE - LTE
11:47:15.392 -> 2: use automatic selection
11:47:15.431 -> 
11:47:15.431 -> Press 1-2 to select an operator.
11:47:20.053 -> Connecting to option 1
11:47:36.147 -> Network vodafone IT registered
11:47:36.147 -> 
11:47:36.147 -> Connection info:
11:47:36.147 -> Context ID:	APN Name:	IP Address:
11:47:36.436 -> 1	ims.mnc010.mcc222.gprs	10.76.21.61
11:47:37.965 -> Operator: 22210
11:47:38.005 -> RSSI: 7
11:47:38.005 -> 
11:47:38.005 -> The available Context IDs are:
11:47:38.005 -> Context ID:	APN Name:	IP Address:
11:47:38.352 -> 1	ims.mnc010.mcc222.gprs	10.76.21.61
11:47:39.770 -> 
11:47:39.770 -> Which Context ID do you want to use for your Packet Switched Data connection?
11:47:39.770 -> Please enter the number (followed by LF / Newline): 
11:47:39.802 -> Using CID: 0
11:47:39.802 -> CCID: 8939104570006261554
11:47:39.889 -> CAPABILITIES: +CGSM
11:47:39.959 -> Warning: performPDPaction (deactivate profile) failed. Probably because no profile was active.
11:47:40.545 -> 
11:47:40.545 -> PSD Action:  result: 0 (success)  IP Address: "10.76.54.73"
11:47:41.532 -> 
11:47:41.532 -> Enter the name of the server you want to ping (followed by LF / Newline): 
11:47:48.837 -> www.google.it
11:47:48.909 -> ++

**Comment by PaulZC** [comment]
Hi David,

I am not sure what to tell you...

The SARA-R510M8S is LTE-M and NB-IoT capable (3GPP Release 13 LTE Cat M1 and NB1).

Using m-center, I can only scan for operators if I have a SIM installed. With my O2 (Telefonica) SIM installed, I see:

![image](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/assets/5690545/da5a779d-cc60-43e9-9e7e-282fef00ba3e)

```vodafone UK``` is new. Previously (last year) I was only able to see ```O2```. Perhaps Vodafone are now also supporting LTE (E-UTRAN)?

My O2 SIM is a standard pay-as-you-go SIM. There's nothing 'special' about it.

Some Googling suggests that Vodafone are the UK's only NB-IoT provider. O2 support LTE-M and "will be rolling out NB-IoT".

More Googling suggests that Vodafone Italy supports LTE-M.

So my advice would be to keep using your Vodafone SIM. It should work. Perhaps there is something in the module configuration which is not correct.

If your SIM is a Vodafone SIM, you should be able to connect OK.

Please check the Context ID, it looks strange:

![image](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/assets/5690545/91a6701e-a41d-4b8d-b96c-8f3b8733d585)

I will repeat your test here. But first I need to buy more credit!

Until later,
Paul

**Comment by PaulZC** [comment]
Here's a successful PING using m-center:

![image](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/assets/5690545/265e26b9-75fe-4bf7-a203-9d82c538114b)

**Comment by PaulZC** [comment]
In m-center, what do you see in the **AT Terminal** when you **Refresh** the **PSD Profiles** on the **Network** tab?

What does ```AT+CGDCONT?``` return?

**Comment by Davefries1966** [comment]
Thanks Paul I will try with m-center, which I already used to scan network and register Vodafone SIM etc. (the same steps as per your examples). I was just wondering how to try a ping from m-center ! :-)

With regards to "strange" CIDS, I simply copied and pasted register+PDP+ping examples into a unique sw code, actually it doesn't let me choose the CID, it keeps on automatically with CID=0; the I tried to force the selection to CID=1, but nothing happens (the actual problem is that the PING callback is never invoked).

I will try the ping with m-center, just to be sure that my SIM works!

**Comment by Davefries1966** [comment]
![image](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/assets/127225643/9174ff63-63de-4209-ae24-a1a8e479d02a)

**Comment by Davefries1966** [comment]
![image](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/assets/127225643/eb2582ba-e95a-4a67-a700-262072391621)

**Comment by Davefries1966** [comment]
![image](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/assets/127225643/3c1c38a5-1677-419e-b314-914e25b22668)

**Comment by Davefries1966** [comment]
The CID number "1" seems to be the correct one. But how can I sure that is the selected CID ?

**Comment by Davefries1966** [comment]
![image](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/assets/127225643/8bb3da6d-91b7-4cea-b8aa-21af1875ddd2)

OK is activated, is the only one that can be deactivated... :-)

**Comment by PaulZC** [comment]
Agreed, CID 1 looks correct. What do you see when you send:

AT+UPSD=0,100,1
AT+UPSD=0,0,2
AT+UPSDA=0,3
AT+UPING="www.google.it"

**Comment by Davefries1966** [comment]
![image](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/assets/127225643/5a9e7456-eddc-4256-929b-a890837e4d56)

**Comment by PaulZC** [comment]
From earlier: Ping error 17 is "PSD or CSD connection not established"
Ping error 8 is "Cannot resolve host"
This sounds like a DNS issue??
Please try AT+UPING="216.58.212.195"

**Comment by PaulZC** [comment]
Please try:
AT+UPSND=0,8
AT+UPSND=0,0

**Comment by Davefries1966** [comment]
![image](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/assets/127225643/f032e5a6-7028-4ff5-8e48-e48b119eadc9)

**Comment by Davefries1966** [comment]
![image](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/assets/127225643/694b017a-ab53-4559-875c-1f229d955a66)

**Comment by PaulZC** [comment]
Please try AT+UPSND=0,0 again
You typed ATAT+UPSND=0,0

**Comment by Davefries1966** [comment]
Sorry...

![image](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/assets/127225643/5cc0dacc-28b1-45a7-85b7-9a44cb121da3)

**Comment by PaulZC** [comment]
Thanks David,

I am out of ideas. I will need to ask a u-blox Field Application Engineer for more suggestions. It will take some time - the ones I know are in the USA...

Until later,
Paul

**Comment by Davefries1966** [comment]
I am quite sure that the DNS is not enabled on the SIM...I am not an expert, but how the cellphone can browse websites without a DNS set ? Or send whatsapp messages ?

**Comment by Davefries1966** [comment]
Paul,

thank you for your patience and support ! I will try to set a DNS on the SIM. 

Cheers
David

**Comment by PaulZC** [comment]
Ah! This might help:

https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/blob/main/src/SparkFun_u-blox_SARA-R5_Arduino_Library.h#L251-L276

**"The standard Europe profile should be used as the basis for all other MNOs in Europe outside of Vodafone and Deutsche Telekom."**

Please try:

```
const mobile_network_operator_t MOBILE_NETWORK_OPERATOR = MNO_VODAFONE;
```

**Comment by Davefries1966** [comment]
OK... anyway the behaviour with m-center is very strange (I also checked the DNS on the SIM, it's correctly set).

**Comment by PaulZC** [comment]
Hi David,

I don't think the firmware version is your issue. I am running 02.05...

The u-blox FAE thinks that you may indeed need a Vodafone IOT SIM. He says that some networks are "picky"; Verizon (USA) only allows specific SIMs on its CAT-M network. I guess I'm lucky - being able to use a standard SIM on O2 (Telefonica).

He also isn't sure where your "lte.postemobile.it" CID is coming from. Is that something you may have entered yourself? He suggests trying:

```
AT+UPSD=0,0,0
AT+UPSD=0,1,"lte.postemobile.it"
AT+UPSD=0,100,2
AT+UPSDA=0,3
```

He also says if you did manually enter it, have you tested with it blank to see what is populated after registration, with CAT-M it is common that the APN is provided by the network (except in some MVNO roaming SIMs)

I hope this helps,
Paul

**Comment by Davefries1966** [comment]
Hi Paul

I think you are very right, unfortunately here in Italy every option is not for free, as for example could be in UK.
Anyway both Vodafone and Tim offer IoT SIM cards at honest price (i.e. 2 euros per month and 200MB).
Tomorrow I will check at some Vodafone and/or Tim reseller.

I would have never arrived at this point without your support.

I THANK YOU AGAIN for your time spent for me. :-)

Cheers
David

**Comment by Davefries1966** [comment]
Hi Paul

I just activated the Hologram SIM which came with the Sparkfun board, with a "pay as you go" plan (very good indeed).

The board now DOES WORK flawlessly ! :-)

Here a snapshot with m-center:

![image](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/assets/127225643/f0482325-e755-4da9-a34e-945cc4890ed5)

**Comment by PaulZC** [comment]
Hi David,

Excellent - thanks for the update!

Paul

**Comment by Davefries1966** [comment]
Hi Paul,

sorry for disturbing you again, I'm running all "Asset Tracker" examples to deeply understand how the board and library work.

In particular, I'm focused on GNSS activation and detection, so I ran the **example nr 27** but I got the following error:

  if (myGNSS.begin() == false) //Connect to the Ublox module using Wire port
  {
    SERIAL_PORT.println(F("**u-blox GPS not detected at default I2C address. Please check wiring. Freezing.**"));
    while (1);
  }

**All the softwares** I tried with communication between board and GNSS module cannot detect the GNSS, even the HPG software which uses a GNSS receiver object as member of the SARA_R5 object.
Any example calling GNSS begin method returns false.

May I do any further check with m-center and/or would you suggest any simple but significant example to run ?
Must the GNSS antenna be under a clean sky to get GNSS module activating ?

Thanks as usual for your help.

Cheers
David

**Comment by PaulZC** [comment]
Hi David,

Are you using the GNSS internal to the SARA-R5? Or a separate GNSS module?

The internal GNSS has an AT command interface which is very different to the standard module UART/I2C interface.
Please see [this example](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/blob/main/examples/SARA-R5_Example1_GNSS_GPRMC/SARA-R5_Example1_GNSS_GPRMC.ino) for more details.

The HPG code uses: LARA-R6 for cellular communication; ZED-F9P for positioning; and a NEO-D9S for L-Band corrections from PointPerfect.

Best wishes,
Paul

**Comment by PaulZC** [comment]
Hi David,

Apologies for the confusion...

For the [SARA-R5 examples](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/tree/main/examples):
- Example1 uses the internal GNSS
- Example12/14/15/16 use an external GNSS
- Any example using ```SparkFun_u-blox_GNSS_Arduino_Library.h``` is using an external GNSS module

For the [Asset Tracker examples](https://github.com/sparkfun/MicroMod_Asset_Tracker/tree/main/Examples):
- Example11/23/24 uses the internal GNSS
- Example25/26/27/28 use an external GNSS
- Any example using ```SparkFun_u-blox_GNSS_Arduino_Library.h``` is using an external GNSS module

For RTK / NTRIP / L-Band, you need a receiver capable of RTK. The dual-band [ZED-F9P](https://www.sparkfun.com/products/16481) is always a good choice. But there are cheaper single-band alternatives: [NEO-M8P](https://www.sparkfun.com/products/15005) plus [u.FL to SMA adapter](https://www.sparkfun.com/products/9145).

The [NEO-D9S](https://www.sparkfun.com/products/19390) is a separate receiver. And it needs an L-Band capable antenna. We also do a [combo board](https://www.sparkfun.com/products/22560) which has the NEO and the ZED on a single board, with antenna sharing.

All of the above will only work correctly if the antenna has a clear view of the sky.

All the best,
Paul

**Comment by Davefries1966** [comment]
Hi Paul

thank you so much for the explanation, now it's very clear to me how to navigate through the examples! :-)

**Comment by Davefries1966** [comment]
Hi Paul,

just for curiosity, would it be possible to use an external data augmentation service (e.g., thingstream PP, NTRIP etc) and send the received correction packets to internal GNSS receiver ? Is there any Asset Tracker example that already does that ?

For example, the HPG solution calls:

**Gnss.inject(msg);**

each time a correction topic is received... how can I do the same towards the internal GNSS receiver ? 

Cheers
David

**Comment by Davefries1966** [comment]
I woud mean that, on one side, I am now able to read MQTT topics (from Thingstream) as MQTT client (via WIFI or LTE), on the other side I can read GNSS GPRMC as per Asset Tracker example 11. How can I provide the correction topics to the internal GNSS receiver to get more precise positions ?

**Comment by PaulZC** [comment]
Hi David,

The internal GNSS is an M8. It is single-band and not RTK-capable. You might get a _small_ improvement if you can figure out how to inject RTCM to it. But it's not going to give you cm accuracy. For that, you need a dual-band RTK-capable GNSS like the ZED-F9P.

Again, if you can figure out how to do it, injecting AssistNow data into the M8 will speed up its Time-To-First-Fix. But it won't help the accuracy.

Cheers,
Paul

**Comment by PaulZC** [comment]
If the accuracy seems really bad, i.e. worse than 10m, the SARA is probably using Cell-Locate instead of GNSS. Cell-Locate gives you a very approximate location based on the signals from the cellular 'towers'.

**Comment by Davefries1966** [comment]
I understand... I should move all my stuff on the top of the buiding where I have a clean sky.
I will let you know about the result. :-)

**Comment by PaulZC** [comment]
Just move the antenna. Keep everything else warm and dry! ;-)

Use good-quality [RG58 cable](https://www.sparkfun.com/products/21281). It's low-loss. RG174 will just leave you wondering where your signal has gone...

**Comment by Davefries1966** [comment]
Hi Paul

I would ask you something about hardware stuff, cause I am now using LTE hinged antenna but I need instead a cabled one (I would mount the Asset Tracker inside a waterproof box with cables exiting from the container).

In your opinion, is the following 4G LTE antenna (12 dBi gain) suitable for the Sparkfun Asset Tracker board ?

https://www.amazon.it/dp/B09WTZ78D9?psc=1&ref_=cm_sw_r_cp_ud_ct_YNW8WFH17TMSF7JEQ731

Or maybe this, lower gain but easier to manage:

https://www.amazon.it/dp/B0BW3MKMPR?psc=1&ref_=cm_sw_r_cp_ud_ct_FZXVEYYVWN85BFHSX05M

As usual, thank you for your response. :-)

Please feel free to suggest any other good antennas which you was friendly with.

Cheers
David

**Comment by PaulZC** [comment]
Hi David,

For indoor work, I use the [SparkFun LTE antenna](https://www.sparkfun.com/products/16432).

But I also have one of these up on the roof: https://www.amazon.co.uk/gp/product/B086J9NBQ8 . It's what I use for most of my SARA-R5 testing. I use RG58 SMA male-female to extend the cable.

Best,
Paul

**Comment by Davefries1966** [comment]
Thanks!!!

**Comment by Davefries1966** [comment]
I still haven't understood why I get first valid fix after a couple of minutes with example 11 (maybe due to GNNS+LTE combination?) and I did receive always a wrong fix (in Italy but with uncertainty 734000) with example 23 (GNSS request with callback). My u-blox antenna now is on the top of the building under a clean sky... :-(

**Comment by Davefries1966** [comment]
![image](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/assets/127225643/25190aae-2318-4000-b905-cc3b66918b1a)

**Comment by PaulZC** [comment]
Hi David,

The decimal places in your latitude and longitude are suspicious. I think something is rounding the position to the nearest 10 arc-minutes:

![image](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/assets/5690545/7d2077e2-ba48-4435-94f9-0b4f98036218)

I suspect this is a cell-locate position with 10 minute resolution. Or something has gone wrong with the location parsing?

Best,
Paul

**Comment by PaulZC** [comment]
Did you run AssetTracker_Example24_GNSS_CellLocate.ino ? Perhaps the module has remembered the cell-locate server settings?

**Comment by Davefries1966** [comment]
I'm using 7 decimals just for monitoring, but what I got from GNSS example 23 is:

42.833301544189454 12.833299636840820

which leads to the same wrong location.

I don't remember which examples I ran, now I'm using the code of example 23

**Comment by Davefries1966** [comment]
maybe may I have to call the method enableGNSSantenna power ?

Please consider that also with m-center I got the same wrong location...maybe I must turn on/off some pins or else ?

**Comment by Davefries1966** [comment]
For sure I ran the example 11 with the following instructions:

 // Set the UTIME mode to pulse-per-second output using a best effort from GNSS **and LTE**
  assetTracker.setUtimeMode(); // Use defaults (mode = SARA_R5_UTIME_MODE_PPS, sensor = **SARA_R5_UTIME_SENSOR_GNSS_LTE**)

with this example I get a quite good positioning after 2 minutes

**Comment by PaulZC** [comment]
Hi David,

Yes, do enable the GNSS antenna power. If you are using an active antenna, you will get almost zero signal if the antenna power is not present.

Also, [gpsRequest](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/blob/bc6c50f7345644aa518315734d47cd3488f90681/src/SparkFun_u-blox_SARA-R5_Arduino_Library.h#L1001) defaults to sensor 3 (1 + 2) which is GNSS plus CellLocate. If you set sensor to 1, it will only use the GNSS.

I hope this helps,
Paul

**Comment by Davefries1966** [comment]
THAT DOES WORK ! :-)

**Comment by Davefries1966** [comment]
Finally I got right position!!

17:15:53.881 -> GNSS Data Received
17:15:53.881 -> ==================
17:15:53.881 -> Date (MM/DD/YYYY): 10/27/2023
17:15:53.881 -> Time: 15:16:08.000
17:15:53.881 -> Lat: 41.9519577
17:15:53.881 -> Lon: 12.7215385
17:15:53.916 -> Alt: 91.00
17:15:53.916 -> Uncertainty: 4
17:15:53.916 -> Speed: 0.00
17:15:53.916 -> Course Over Ground: 0.00

**Comment by Davefries1966** [comment]
It's exactly where I am... great!!!!

![image](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/assets/127225643/bc5d183e-f4a5-4b35-be98-90e11c43be10)

**Comment by brhoff720** [comment]
Hey Paul, can this be closed? Do we need Docs to update anything.

**Comment by Davefries1966** [comment]
Hi all, for me the issue can be closed. I finally succeeded in realizing my prototype, thanks to the great helps from Paul. Without him, the project in figures below would have never seen the light. :-)

![cat6](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/assets/127225643/9dbb5895-69d5-4bc0-b78c-e18ee6da188b)
![cat5](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/assets/127225643/412e910c-17fe-477e-b618-aa4aafe284f4)
![cat4](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/assets/127225643/381f7c5e-65ce-4a9e-9af6-a9bb5f7a405b)
![cat3](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/assets/127225643/75c07e83-7ec3-44dc-ac7b-509f4e059122)
![cat2](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/assets/127225643/02438451-e561-42be-8f8a-f1497916925a)
![cat1](https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/assets/127225643/5e1a7234-3c79-40bf-8cc1-c62244d95c18)

**Comment by PaulZC** [comment]
Hi,

@Davefries1966  - thanks for the update and for sharing the photos. That looks like an excellent project! I'm glad everything is working now.

@brhoff720 : 

Can this be closed? No, not yet. I'm still keen to write an Assist-Now via MQTT example for the SARA-R5 library, running on the Asset Tracker. I think I've figured out how to push the assist now data to the SARA's internal GNSS, so it would be a good example to demonstrate: how to use MQTT on the SARA; and how Assist Now speeds up the Time-To-First-Fix. (The MQTT part would also be applicable to the LARA, but not the internal-GNSS part as the LARA does not have one.)

Do Docs need to update anything? Yes. The above discussion with David was useful. There are a few things that could be clearer in our documentation and hook-up guide:
Hologram works well in Italy. David wasn't able to connect to Vodafone (LTE), but that may have been a SIM issue.
We could be clearer about which examples use the internal GNSS and which an external GNSS: https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/issues/34#issuecomment-1778906064
We could be clearer about the need to enable power for an active GNSS antenna: https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/issues/34#issuecomment-1783063475

Cheers,
Paul

**Comment by Davefries1966** [comment]
Thank you all guys, you are very great! :-)

I'm looking forward to further example about Assist-Now via MQTT, it will surely improve my project.

Cheers
David

**Comment by PaulZC** [comment]
Hi David (@Davefries1966 ),

Please see https://github.com/sparkfun/SparkFun_u-blox_SARA-R5_Arduino_Library/releases/tag/v1.1.9 for the new MQTT example.

Enjoy!

Best wishes,
Paul

**Comment by PaulZC** [comment]
You are welcome David,

I've kept that example as simple as possible. It publishes data to a single channel. And subscribes to that same single channel (topic).

If you are subscribed to multiple topics:
* In ```processMQTTcommandResult```, look for ```command == SARA_R5_MQTT_COMMAND_READ``` (6). ```result``` contains the number of unread messages.
* Call ```readMQTT``` multiple times to read all the messages. ```topic``` contains the topic; ```buf``` contains the actual data.

Best,
Paul

**Comment by Davefries1966** [comment]
Hi Paul,

your example is very clear, and the suggestion about the "result" is really useful.
Thank you for all !

Cheers
David

**Comment by tyrian869** [comment]
Hello I am trying to run SARA-R5-Example2 with my SARA-R510M8S that has a 600mhz-6000mhz low profile antenna attached to it,  and per the serial monitor my SIM I'm getting the bellow information, does it look correct or am I supposed to get more sim state report information?

I'm using hologram SIM that supports two network providers with LTE-M / CAT-M1 Bands. what my coverage says for each of the two networks:

4G LTE LTE-M NB-IoT 3G 2G

Following the Sparkfun guide I don't ever see a blue light appear on my SARA-R5 (NI) led, I only have a bright red for (ON) and a somewhat bright but more dim led for the (3.3) led.

My sim is detected as well as the Sara-R5 information but it never moves past sim state reports? 

All X's are populated just renamed for security reason. 







SARA-R5 connected!

Manufacturer ID:     u-blox
Model ID:            SARA-R510M8S
Firmware Version:    03.30
Product Serial No.:  XXXXXXXXXXXXXX
IMEI:                XXXXXXXXXXXXX
IMSI:                XXXXXXXXXXXXX
SIM CCID:            XXXXXXXXXXXXXXXX

Subscriber No.:      
Capabilities:        +CGSM
SIM state reports requested...

**Comment by PaulZC** [comment]
Hi @tyrian869 ,

Please avoid cross-posting. I see you have asked the same question on our forum:

https://community.sparkfun.com/t/sara-r5-function-board-appears-to-hang-after-an-unpredictable-amount-of-time/46828/5

We will try and help you there.

Best wishes,
Paul
